File: js/medium/longest_substring.js
const lengthOfLongestSubstring = (s) => {
  left_p = 0;
  right_p = 0;
  max = 0;
  let hash_set = new Set();
  while (right_p < s.length) {
    if (!hash_set.has(s[right_p])) {
      hash_set.add(s[right_p]);
      max = Math.max(max, hash_set.size);
      right_p++;
    } else {
      hash_set.delete(s[left_p]);
      left_p++;
    }
  }
  return max;
};

console.log(lengthOfLongestSubstring("abcabcbb"));

File: py/medium/longest_substring.py

def lengthOfLongestSubstring(s: str) -> int:
    left_pointer = 0
    right_pointer = 0
    max_size = 0
    hash_set = set()
    while(right_pointer < len(s)):
        if (s[right_pointer] not in hash_set):
            hash_set.add(s[right_pointer])
            right_pointer += 1
            max_size = max(len(hash_set), max_size)
            print(hash_set, max_size)
        else:
            hash_set.remove(s[left_pointer])
            left_pointer += 1
        print(hash_set, max_size, 'here')
        # right_pointer += 1 never increment right pointer outside if it will cause left pointer to
        # lag behind

    return max_size

print(lengthOfLongestSubstring('abcabcbb'))